test/integration/vhds_integration_test.cc

```diff
@@ -653,5 +653,40 @@ TEST_P(VhdsIntegrationTest, VhdsOnDemandUpdateFailToResolveOneAliasOutOfSeveral)
   cleanupUpstreamAndDownstream();
 }
 
+// Verify that an vhds update succeeds even when the client closes its connection
+TEST_P(VhdsIntegrationTest, VhdsOnDemandUpdateHttpConnectionCloses) {
+  // RDS exchange with a non-empty virtual_hosts field
+  useRdsWithVhosts();
+
+  testRouterHeaderOnlyRequestAndResponse(nullptr, 1);
+  cleanupUpstreamAndDownstream();
+  codec_client_->waitForDisconnect();
+  Stats::Store& stats = test_server_->server().stats();
+  const auto initial_config_reloads =
+      TestUtility::findCounter(stats, "http.config_test.vhds.my_route.config_reload")->value();
+
+  // Attempt to make a request to an unknown host
+  codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
+  Http::TestRequestHeaderMapImpl request_headers{{":method", "GET"},
+                                                 {":path", "/"},
+                                                 {":scheme", "http"},
+                                                 {":authority", "vhost_1"},
+                                                 {"x-lyft-user-id", "123"}};
+  IntegrationStreamDecoderPtr response = codec_client_->makeHeaderOnlyRequest(request_headers);
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().VirtualHost,
+                                           {vhdsRequestResourceName("vhost_1")}, {}, vhds_stream_));
+
+  envoy::api::v2::DeltaDiscoveryResponse vhds_update =
+      createDeltaDiscoveryResponseWithResourceNameUsedAsAlias();
+  vhds_stream_->sendGrpcMessage(vhds_update);
+
+  codec_client_->close();
```

Comment:
So I think a more interesting test would be to reset the stream, but not the connection.  In that case the HCM will persist beyond the life of the stream, and I think you'll hit the lifetime issue Matt and I see.

I think rather than a simple lambda, you need something with a bit more state, so you can cancel the actual callback if the filter goes away.

Reply:
> I think rather than a simple lambda, you need something with a bit more state, so you can cancel the actual callback if the filter goes away.

The state of the callback (whether it and the filter-chain are still active) is handled implicitly via shared_ptr - weak_ptr combination (pls. see https://github.com/envoyproxy/envoy/issues/9784#issuecomment-635034755 for more details)

Reply:
If it's not clear in the code, (which IMO it's not, especially given how many autos there are in here:
https://github.com/envoyproxy/envoy/blob/master/source/common/router/rds_impl.cc#L288
making it hard to tell what's going on w.r.t. weak pointers - mind adding some comments to onDestroy to explain it for posterity?

Also, please change this test to reset the stream without closing the connection.
thanks!